snapcraft_legacy/storeapi/v2/validation_sets.py

```diff
@@ -14,161 +14,130 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from typing import Any, Dict, List, Optional
+from typing import Any, Annotated, Dict, Literal, TYPE_CHECKING
+import pydantic
 
-import jsonschema
+from craft_application import models
 
-from ._api_schema import BUILD_ASSERTION_JSONSCHEMA
 
+if TYPE_CHECKING:
+    SnapName = str
+    SnapId = str
+else:
+    SnapName = pydantic.constr(max_length=40)
+    SnapId = pydantic.constr(max_length=40)
 
-class Snap:
+
+def cast_dict_scalars_to_strings(data: dict) -> dict[str, Any]:
+    """Cast all scalars in a dictionary to strings.
+
+    Supported scalar values are str, int, float, and bool.
+    """
+    return {_to_string(key): _to_string(value) for key, value in data.items()}
+
+
+def _to_string(
+    data: dict | list | str | int | float | str | bool | None
+) -> dict[str, Any] | list | str | None:
+    """Recurse through nested dicts and lists and cast scalar values to strings.
+
+    Supported scalar values are str, int, float, and bool.
+    """
+    if isinstance(data, dict):
```

Comment:
It might be worth starting out with the trivial case for performance:

```suggestion
    if isinstance(data, str):
        return data

    if isinstance(data, dict):
```